### ISSUE TITLE
Introduce `types::fe_index`.

### ISSUE THREAD

**Issue by marcfehling** [dealii/dealii#13595]
Currently, `active_fe_index` and `future_fe_index` instances are of `unsigned int` type (32-bit). I don't imagine that we will ever use up to 4 billion different finite elements in a `hp::FECollection` object, so I would propose to use a `char` type (8-bit) in the long run.

To give us more flexibility in choosing datatypes, I would suggest to first introduce `types::fe_index`. We can refactor the library next to work with this new type, and then move to a smaller datatype `unsigned char` (or any other datatype) in a last step.

While writing checkpoints to the filesystem for my hp-experiments, a serialized container for the `active_fe_indices` takes up as much space as the whole dump of `p4est` (~1 billion cells: 880 MB `p4est`, 590 MB `active_fe_indices`). Depending on the ZLIB compression, we can reduce the size of the serialized container up to a factor of 4.

With this change, we can also reduce the memory demand of the `fe_indices` at runtime. However, the `fe_index` part never consumed much memory in the first place compared to the other parts of a deal.II program. So this suggestion must therefore be understood as purely cosmetic, rather than fixing something critical. So this is no solution to #9598.

EDIT: Actually, we use `unsigned short int` as a type for FE indices internally in the `DoFHandler` class. But the whole interface in accessing and setting indices uses `unsigned int`. We should move the type declaration into the global scope and clean this up!

**Comment by drwells** [comment]
If we are using inconsistent types in the library then that's a bug - we should clean that up regardless. In the past, though, we have tried to avoid using `unsigned char` et al in public interfaces because it is annoying to print.

Could we instead internally convert back-and-forth between `unsigned char` and `unsigned short int` during serialization?

**Comment by marcfehling** [comment]
`uint8_t` or `uint16_t` would also be possible data types. But choosing one would be the very last step :)

**Comment by bangerth** [comment]
That ship has long sailed -- we're using `std::uint*_t` all over the place already :-)

I'm in favor of `uint*_t` over `unsigned char` and similar because the former clearly indicates that it is an integer type, whereas the latter suggests an abused character type. They are both equally annoying to print when interpreted as integers...

**Comment by drwells** [comment]
You're right - my recollections of these topics have some holes!

**Comment by marcfehling** [comment]
We will also need to bump the serialization version, as we will use a different datatype for active FE index checkpoints.

https://github.com/dealii/dealii/blob/e6779913e7f770652da442b3c2460dec3e864029/source/distributed/tria.cc#L2024

**Comment by marcfehling** [comment]
I've been working on this recently to create a coherent interface. I realize that the collection objects in the `hp` namespace can potentially contain more elements than `active_fe_index` allows us to address.

I would like to limit the size of at least the `FECollection` object. We can leave `QCollection` and `MappingCollection` etc untouched for now.

Do you think we should extend this limit on all `Collection` objects?

**Comment by marcfehling** [comment]
Until today, we implemented the new `types::fe_index` datatype in `DoFHandler` and all related accessor classes, or in other words, in all data structures within the `dofs/` folder.

The question remains if it make sense to continue to adapt the new datatype to all other parts of the library. I came to the conclustion that (a) other classes are not dependent on the choice of FE index datatype, (b) it doesn't really matter memory-wise as other classes will not likely store a large amount of FE indices, but rather just use one temporary index for their internal operations, and (c) it is not worth the effort and time to adjust the datatype in the whole library. For example, `FECollection` itself does not need to be limited to contain a number of elements only representable by 8- or 16-bit integers. This requirement only comes from the `DoFHandler` implementation, and thus we should only limit it there.

Now, we use 16-bit unsigned integers (`unsigned short int`) as `types::fe_index`. We debated about using 8-bit types instead (`unsigned char`). The only problem is that printing this datatype using e.g. the stream operator `operator<<` will interpret these numbers as characters and will produce weird or no output at all. To avoid confusion and hacking around the problem, let's settle with 16-bit integers. We do not waste a lot of memory with FE indices anyways.

I will this close this issue as completed.
